### PR TITLE
[Test] Add test to check dashboard URL

### PR DESCRIPTION
**Pull request by sayaleeraut**
Test case covered - CEPH-83575611
Test steps -
1) Deploy ceph cluster with dashboard
2) Perform mgr container restart and mgr service failure 3) Verify the dashboard service does not bind to
host.containers.internal

# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

File: tests/mgr/test_dashboard_url.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    # Get dashboard URL from mgr service
+    url = ceph_cluster.get_mgr_services().get("dashboard")
+    if not url:
+        raise ResourceNotFoundError("Dashboard not configured")
+
+    # Check if dashboard URL does not bind to host.containers.internal
+    if "host.containers.internal" in url:
+        raise DashboardUrlError("Dashboard URL binds to host.containers.internal")
+    log.info("Dashboard URL is {}".format(url))
```
Comment (pranavprakash20) [inline tests/mgr/test_dashboard_url.py:73]: One suggestion Sayalee, can we move this piece of code

```
url = ceph_cluster.get_mgr_services().get("dashboard")
    if not url:
        raise ResourceNotFoundError("Dashboard not configured")

    # Check if dashboard URL does not bind to host.containers.internal
    if "host.containers.internal" in url:
        raise DashboardUrlError("Dashboard URL binds to host.containers.internal")
    log.info("Dashboard URL is {}".format(url))```
    
    to a method named 'check_dashboard' or 'validate_dashboard_url' etc and call it from run? As I can see this piece of code is called 3 times inside the run method.

**Review by pranavprakash20 — COMMENTED** [review]
One suggestion added, please check

File: tests/mgr/test_dashboard_url.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    out, rc = node.exec_command(sudo=True, cmd="podman ps | grep mgr")
+    container_id = out.split(" ")[0]
+
+    # Restart container for mgr daemon
+    out, rc = node.exec_command(sudo=True, cmd=f"podman restart {container_id}")
+
+    # Get dashboard URL from mgr service
+    url = ceph_cluster.get_mgr_services().get("dashboard")
+    if not url:
+        raise ResourceNotFoundError("Dashboard not configured")
```
Comment (adityaramteke) [inline tests/mgr/test_dashboard_url.py:49]: @sayaleeraut why do we have to check the dashboard URL twice?

Comment (sayaleeraut) [reply tests/mgr/test_dashboard_url.py:49]: Hey @adityaramteke Do you mean after container restart and mgr fail respectively ??

File: tests/mgr/test_dashboard_url.py (hunk 1 above)
Comment (sayaleeraut) [reply tests/mgr/test_dashboard_url.py:73]: Sure @pranavprakash20 Will add the changes

File: tests/mgr/test_dashboard_url.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    Get dashboard URL from mgr service and check if dashboard URL does not
+    bind to host.containers.internal
+    """
+    url = ceph_cluster.get_mgr_services().get("dashboard")
+    if not url:
+        raise ResourceNotFoundError("Dashboard not configured")
+    if "host.containers.internal" in url:
+        log.error("Dashboard URL {} is not binding to IP of hostname")
+        return False
+    log.info("Dashboard URL is {}".format(url))
```
Comment (vamahaja) [inline tests/mgr/test_dashboard_url.py:31]: Use `f"Dashboard URL is {url}"`

File: tests/mgr/test_dashboard_url.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+def check_dashboard_url(ceph_cluster):
+    """
+    Get dashboard URL from mgr service and check if dashboard URL does not
+    bind to host.containers.internal
+    """
+    url = ceph_cluster.get_mgr_services().get("dashboard")
+    if not url:
+        raise ResourceNotFoundError("Dashboard not configured")
+    if "host.containers.internal" in url:
+        log.error("Dashboard URL {} is not binding to IP of hostname")
```
Comment (vamahaja) [inline tests/mgr/test_dashboard_url.py:29]: missing `url` to pass  to message?

File: tests/mgr/test_dashboard_url.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+class DashboardUrlError(Exception):
+    pass
+
+
+class CephMgrServiceFailError(Exception):
+    pass
+
+
+class ResourceNotFoundError(Exception):
```
Comment (vamahaja) [inline tests/mgr/test_dashboard_url.py:16]: Instead use - https://github.com/red-hat-storage/cephci/blob/master/cli/exceptions.py#L20

File: tests/mgr/test_dashboard_url.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+from utility.log import Log
+
+log = Log(__name__)
+
+
+class DashboardUrlError(Exception):
+    pass
+
+
+class CephMgrServiceFailError(Exception):
```
Comment (vamahaja) [inline tests/mgr/test_dashboard_url.py:12]: instead use - https://github.com/red-hat-storage/cephci/blob/master/cli/exceptions.py#L32

File: tests/mgr/test_dashboard_url.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
+from cli.cephadm.cephadm import CephAdm
+from cli.utilities.utils import get_service_id
+from utility.log import Log
+
+log = Log(__name__)
+
+
+class DashboardUrlError(Exception):
```
Comment (vamahaja) [inline tests/mgr/test_dashboard_url.py:8]: This is not required.

File: tests/mgr/test_dashboard_url.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    Check if the Ceph dashboard URL consists of host IP and
+    does not bind to host.containers.internal before and after
+    container restart and mgr failover is performed
+    """
+    # Get mgr node
+    node = ceph_cluster.get_nodes(role="mgr")[0]
+
+    # Check if dashboard URL does not bind to host.containers.internal
+    check_url = check_dashboard_url(ceph_cluster)
+    if check_url is False:
```
Comment (vamahaja) [inline tests/mgr/test_dashboard_url.py:46]: Unnecessary check

File: tests/mgr/test_dashboard_url.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    # Get container id for mgr daemon
+    out, rc = node.exec_command(sudo=True, cmd="podman ps | grep mgr")
+    container_id = out.split(" ")[0]
+
+    # Restart container for mgr daemon
+    out, rc = node.exec_command(sudo=True, cmd=f"podman restart {container_id}")
+
+    # Check if dashboard URL does not bind to host.containers.internal
+    check_url = check_dashboard_url(ceph_cluster)
+    if check_url is False:
```
Comment (vamahaja) [inline tests/mgr/test_dashboard_url.py:59]: Unnecessary check

File: tests/mgr/test_dashboard_url.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+    # Fail mgr service
+    mgr_service = get_service_id(node, "mgr")[0].split("@")[1][4:-8]
+    out = CephAdm(node).ceph.mgr.fail(mgr=mgr_service)
+    if out:
+        raise CephMgrServiceFailError("Failed to execute `ceph mgr fail` command")
+
+    # Check if dashboard URL does not bind to host.containers.internal
+    check_url = check_dashboard_url(ceph_cluster)
+    if check_url is False:
```
Comment (vamahaja) [inline tests/mgr/test_dashboard_url.py:74]: Unnecessary check

File: tests/mgr/test_dashboard_url.py [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    node = ceph_cluster.get_nodes(role="mgr")[0]
+
+    # Check if dashboard URL does not bind to host.containers.internal
+    check_url = check_dashboard_url(ceph_cluster)
+    if check_url is False:
+        raise DashboardUrlError("Dashboard URL binds to host.containers.internal")
+    log.info("Dashboard URL does NOT bind to host.containers.internal")
+
+    # Get container id for mgr daemon
+    out, rc = node.exec_command(sudo=True, cmd="podman ps | grep mgr")
```
Comment (vamahaja) [inline tests/mgr/test_dashboard_url.py:51]: instead use - https://github.com/red-hat-storage/cephci/blob/master/cli/utilities/utils.py#L60

File: tests/mgr/test_dashboard_url.py [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    container_id = out.split(" ")[0]
+
+    # Restart container for mgr daemon
+    out, rc = node.exec_command(sudo=True, cmd=f"podman restart {container_id}")
```
Comment (vamahaja) [inline tests/mgr/test_dashboard_url.py:55]: instead add this method here - https://github.com/red-hat-storage/cephci/blob/master/cli/utilities/utils.py#L100

File: tests/mgr/test_dashboard_url.py [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    if check_url is False:
+        raise DashboardUrlError("Dashboard URL binds to host.containers.internal")
+    log.info("Dashboard URL does NOT bind to host.containers.internal")
+
+    # Refresh the ceph orch ps command
+    CephAdm(node).ceph.orch.ps(refresh="True")
+
+    # Fail mgr service
+    mgr_service = get_service_id(node, "mgr")[0].split("@")[1][4:-8]
+    out = CephAdm(node).ceph.mgr.fail(mgr=mgr_service)
```
Comment (vamahaja) [inline tests/mgr/test_dashboard_url.py:68]: better to have regex

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: tests/mgr/test_dashboard_url.py (hunk 11 above)
Comment (sayaleeraut) [reply tests/mgr/test_dashboard_url.py:51]: Hey @vamahaja , If I use get_running_containers() method, the output looks as below-
`[
  {
    "AutoRemove": true,
    "Command": [
      "-n",
      "mgr.ceph-saya-patch-yf9uqs-node1-installer.hzpdiz",
      "-f",
      "--setuser",
      "ceph",
      "--setgroup",
      "ceph",
      "--default-log-to-file=false",
      "--default-log-to-journald=true",
      "--default-log-to-stderr=false"
    ],
    "CreatedAt": "27 hours ago",
    "Exited": false,
    "ExitedAt": -62135596800,
    "ExitCode": 0,
    "Id": "12291a94af7435b40ea70bbe93a65b02a6603a441645c9bb0d05de12595eafaa",
    "Image": "registry-proxy.engineering.redhat.com/rh-osbs/rhceph:ceph-6.1-rhel-9-containers-candidate-42152-20230713054431",
    "ImageID": "85cb9476225e77839acf93d726ab3f510398d978d4741d41e846d10ecab19b4e",
    "IsInfra": false,
    "Labels": {
      "CEPH_POINT_RELEASE": "",
      "GIT_BRANCH": "main",
      "GIT_CLEAN": "True",
      "GIT_COMMIT": "0727c855af939c6f3709e73be703026388413744",
      "GIT_REPO": "https://github.com/ceph/ceph-container.git",
      "RELEASE": "main",
      "architecture": "x86_64",
      "build-date": "2023-07-13T05:44:32",
      "ceph": "True",
      "com.redhat.component": "rhceph-container",
      "com.redhat.license_terms": "https://www.redhat.com/agreements",
      "description": "Red Hat Ceph Storage 6",
      "distribution-scope": "public",
      "io.buildah.version": "1.29.0",
      "io.k8s.description": "Red Hat Ceph Storage 6",
      "io.k8s.display-name": "Red Hat Ceph Storage 6 on RHEL 9",
      "io.openshift.expose-services": "",
      "io.openshift.tags": "rhceph ceph",
      "maintainer": "Guillaume Abrioux \u003cgabrioux@redhat.com\u003e",
      "name": "rhceph",
      "release": "192",
      "summary": "Provides the latest Red Hat Ceph Storage 6 on RHEL 9 in a fully featured and supported base image.",
      "url": "https://access.redhat.com/containers/#/registry.access.redhat.com/rhceph/images/6-192",
      "vcs-ref": "dec93361f6f7a22d929d690d9002f0df9a8f6805",
      "vcs-type": "git",
      "vendor": "Red Hat, Inc.",
      "version": "6"
    },
    "Mounts": [
      "/var/lib/ceph/mgr/ceph-ceph-saya-patch-yf9uqs-node1-installer.hzpdiz",
      "/etc/ceph/ceph.conf",
      "/etc/hosts",
      "/var/run/ceph",
      "/var/log/ceph",
      "/var/lib/ceph/crash",
      "/run/systemd/journal"
    ],
    "Names": [
      "ceph-a18090bc-2152-11ee-a7d6-fa163e318b7b-mgr-ceph-saya-patch-yf9uqs-node1-installer-hzpdiz"
    ],
    "Namespaces": {

    },
    "Networks": [],
    "Pid": 1418733,
    "Pod": "",
    "PodName": "",
    "Ports": null,
    "Size": null,
    "StartedAt": 1689584695,
    "State": "running",
    "Status": "Up 27 hours",
    "Created": 1689584695
  }
]`

It does not contain the "container id" which is present in output of `podman ps | grep mgr command`
 -
 
 # podman ps | grep mgr
**12291a94af74**  registry-proxy.engineering.redhat.com/rh-osbs/rhceph:ceph-6.1-rhel-9-containers-candidate-42152-20230713054431                -n mgr.ceph-saya-...  27 hours ago  Up 27 hours              ceph-a18090bc-2152-11ee-a7d6-fa163e318b7b-mgr-ceph-saya-patch-yf9uqs-node1-installer-hzpdiz

File: tests/mgr/test_dashboard_url.py (hunk 13 above)
Comment (sayaleeraut) [reply tests/mgr/test_dashboard_url.py:57]: Done, added the change

File: tests/mgr/test_dashboard_url.py (hunk 12 above)
Comment (sayaleeraut) [reply tests/mgr/test_dashboard_url.py]: Done

File: tests/mgr/test_dashboard_url.py (hunk 10 above)
Comment (sayaleeraut) [reply tests/mgr/test_dashboard_url.py]: Done, removed it.

File: tests/mgr/test_dashboard_url.py (hunk 9 above)
Comment (sayaleeraut) [reply tests/mgr/test_dashboard_url.py]: Done, removed it.

File: tests/mgr/test_dashboard_url.py (hunk 7 above)
Comment (sayaleeraut) [reply tests/mgr/test_dashboard_url.py]: Done, removed it.

File: tests/mgr/test_dashboard_url.py (hunk 6 above)
Comment (sayaleeraut) [reply tests/mgr/test_dashboard_url.py]: Done, added it.

File: tests/mgr/test_dashboard_url.py (hunk 4 above)
Comment (sayaleeraut) [reply tests/mgr/test_dashboard_url.py]: Done, fixed it

File: tests/mgr/test_dashboard_url.py (hunk 3 above)
Comment (sayaleeraut) [reply tests/mgr/test_dashboard_url.py]: Done, fixed it.

File: tests/mgr/test_dashboard_url.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+
+def check_dashboard_url(ceph_cluster):
+    """
+    Get dashboard URL from mgr service and check if dashboard URL does not
+    bind to host.containers.internal
+    """
+    url = ceph_cluster.get_mgr_services().get("dashboard")
+    if not url:
+        raise ResourceNotFoundError("Dashboard not configured")
```
Comment (vamahaja) [inline tests/mgr/test_dashboard_url.py:18]: Add extra line

File: tests/mgr/test_dashboard_url.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    # Get container id for mgr daemon
+    out, rc = node.exec_command(sudo=True, cmd="podman ps | grep mgr")
+    container_id = out.split(" ")[0]
+
+    # Restart container for mgr daemon
+    restart_container(node, container_id)
+
+    # Check if dashboard URL does not bind to host.containers.internal
+    check_dashboard_url(ceph_cluster)
+    log.info("Dashboard URL does NOT bind to host.containers.internal")
```
Comment (vamahaja) [inline tests/mgr/test_dashboard_url.py:48]: `not`

**Review by vamahaja — CHANGES_REQUESTED** [review]

**Review by pranavprakash20 — APPROVED** [review]
LGTM

File: tests/mgr/test_dashboard_url.py (hunk 14 above)
Comment (sayaleeraut) [reply tests/mgr/test_dashboard_url.py:22]: I have added extra line, Line. no 23

File: tests/mgr/test_dashboard_url.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    """
+    Get dashboard URL from mgr service and check if dashboard URL does not
+    bind to host.containers.internal
+    """
+    url = ceph_cluster.get_mgr_services().get("dashboard")
+    if not url:
+        raise ResourceNotFoundError("Dashboard not configured")
+    if "host.containers.internal" in url:
+        log.error(f"Dashboard URL {url} is not binding to IP of hostname")
+        return False
```
[the remark this answers is not in the document]
Comment (sayaleeraut) [reply tests/mgr/test_dashboard_url.py:26]: I have added extra line, Line. no 27

**Review by vamahaja — APPROVED** [review]

**Pull request merged**